cycle-time/Main.hs

```diff
@@ -0,0 +1,107 @@
+-- | A tool for measuring cycle time of tasks
+--
+-- Cycle time is the difference between creation date and completion date.
+--
+module Main (main) where
+
+import RIO
+
+import Asana.Api
+import Asana.Api.Gid (gidToText, textToGid)
+import Asana.Api.Project (Project(pCreatedAt, pGid, pName), getProjects)
+import Asana.App (loadApp, runApp)
+import Control.Monad (when)
+import Data.Csv as Csv
+import Data.Foldable (maximum, minimum)
+import Data.List (intercalate, nub)
+import qualified Data.Vector as V
+import RIO.ByteString (writeFile)
+import RIO.ByteString.Lazy (toStrict)
+import RIO.Text (isPrefixOf)
+import RIO.Time (NominalDiffTime, diffUTCTime, toGregorian, utctDay)
+import Statistics.Quantile (median, medianUnbiased)
+import Statistics.Sample (kurtosis, mean, skewness, stdDev)
+import Statistics.Sample.Histogram (histogram)
+import System.IO.Temp (emptySystemTempFile)
+import Text.Printf (printf)
+
+main :: IO ()
+main = do
+  app <- loadApp
+  runApp app $ do
+    logDebug "Fetch projects"
+    projects <- filter approvedProject <$> getProjects
+
+    taskGids <-
+      fmap (nub . concat)
+      . pooledForConcurrentlyN maxRequests projects
+      $ \project -> do
+          logDebug . fromString $ "Project tasks: " <> show
+            (gidToText $ pGid project)
+          fmap nGid <$> getProjectTasks (pGid project) AllTasks
+    bugTaskGids <-
+      fmap nGid <$> getProjectTasks (textToGid "56224185180675") AllTasks
+
+    let nonBugGids = filter (`notElem` bugTaskGids) taskGids
+
+    cycleTimes <-
+      zscoreFilter
```

Comment:
Whoa.

Reply:
This is a pretty cool technique for removing extreme outliers.